Org users state module with test coverage

jain-prerna opened the pull request:
Current PR is dependent on [PR-221](https://github.com/saltstack/salt-ext-modules-vmware/pull/221)

As PR-221 has execution module and corresponding test coverage for org_users
and current PR has state module and corresponding test coverage for org_users

jain-prerna commented on tests/integration/states/test_vmc_org_users.py:
```diff
@@ -0,0 +1,54 @@
+"""
```
As we can not expose vmware specific data, we are trying to add a testuser.
The underlying invite API checks if the passed user is part of vmware Active Directory or not. If the user is part of it, then only the underlying invite API will add the user to the given org.

As the testuser is some dummy value and is not part of the vmware Active Directory, the invite API just sends 200 OK response though it did not added the user to the org.

so we are testing the flow as:

In present function, add the user (as the ""test@vmware.com" is not part of vmware Active Directory, the underlying API just sends 200 OK response)
so when we call absent function by passing the name as same ""test@vmware.com", it should return "No user found"

waynew reviewed (commented):
**question** why is the proxy/esxi.py file deleted here?

jain-prerna commented:
By mistake deleted esxi.py earlier, now i have added it back.

waynew commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+        if org_user:
+            return vmc_state._create_state_response(
+                name=name,
+                comment="State absent will remove user with username {}".format(username),
```
**suggestion** reword:

```suggestion
                comment="Would have removed user with username {}".format(username),
```

jain-prerna replied:
Updated

waynew commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+        else:
+            return vmc_state._create_state_response(
+                name=name,
+                comment="State absent will do nothing as no user found with username {}".format(
```
**suggestion** the `test` mode should only be different when salt *would* have changed anything. This should be rewritten:

```
if org_user:
    if __opts__.get("test"):
        # from line 374
        ...
    else:
        # from line 387
        removed_org_user = ...
        ...
else:
    # from line 409
    log.info("No user found ...")
    return ...
```

jain-prerna replied:
If we test with valid email then all the use cases will be covered.
When we are testing with test account (for ex. "test@vmware.com"), it will always return "State absent will do nothing as no user found with username..." in the "test mode" because this test account is not part of vmware active directory and hence it would not be get added to the org while running "invite state" on this account.

waynew replied:
The problem isn't covering all of the cases it's that some of the messaging is **wrong**.

Cases:
 
- test mode
   - user found (A)
   - no user found (B)
- normal/not test mode
   - user found (C)
   - no user found (D)
   
But that's slightly incorrect because B and D _are the same case_. If there is no user found then it doesn't matter. We won't delete the user, nor would we really delete the user.

To see an example:

    salt-call --local state.single user.absent fizzygig test=True

Try this both with `test=True` and without.  There is zero difference in the message. We should do the same thing here:

```
if org_user:
    if __opts__.get('test'):
        # would have removed user (A)
    else:
        # actually removed user (C)
else:        
    # no user found (B & D)
```

jain-prerna replied:
Updated

waynew commented on tests/integration/states/test_vmc_org_users.py:
```diff
@@ -0,0 +1,54 @@
+    data = vmc_connect.copy()
+    data.pop("vcenter_hostname")
+    data.pop("sddc_id")
+    data["hostname"] = data["authorization_host"]
+    data.pop("authorization_host")
```
**suggestion** pop returns the value, this can be one line:
```suggestion
    data["hostname"] = data.pop("authorization_host")
```

jain-prerna replied:
Updated

waynew reviewed (changes requested):
Several suggestions here

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+"""
+VMC Org Users state module
```
make sense, changed it to vmc_org_user

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+           salt-call --local -c /Users/Documents/salt_Configuration/local/etc/salt/ state.single vmc_org_users.absent name="test@vmware.com" hostname=console-stg.cloud.vmware.com org_id=10e1092f-51d0-473a-80f8-137652fd0c39 refresh_key=refresh_key_value verify_ssl=False
+
+    name
+        Indicates the email Id identifying the user to be removed from the organization.
```
Updated

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+
+      .. code-block:: bash
+
+           salt-call --local -c /Users/Documents/salt_Configuration/local/etc/salt/ state.single vmc_org_users.present name="test@vmware.com" hostname=console-stg.cloud.vmware.com org_id=10e1092f-51d0-473a-80f8-137652fd0c39 refresh_key=refresh_key_value organization_roles='[{"name": "org_member"}, {"name": "developer"}]' verify_ssl=False
```
Updated

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+        Its mandatory while sending invitation to user.
+        BaseRoleBinding defines the base binding properties of the role.
+
+        'name': String
+            The name of the role.
+
+        'createdDate': (String) (Optional)
+            The timestamp the role was created at (measured in number of seconds since 1/1/1970 UTC).
+
+        'expiresAt': (Integer As Int64) (Optional)
+            The timestamp the role expires at (measured in number of seconds since 1/1/1970 UTC). Example:3609941597
```
There is provision to create roles this way, so we reaplicated all input parameters from the Cloud Services Platform(CSP) apis documentation

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+        Indicates the email Id identifying the user to be removed from the organization.
+
+    hostname
+        The host name of CSP.
```
Updated with Cloud Services Platform

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+        (Optional) Path to the SSL client certificate file to connect to VMC Cloud Console.
+        The certificate can be retrieved from browser.
+
+    For example:
+
+        .. code::
```
Updated

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+log = logging.getLogger(__name__)
+
+
+def present(
```
Here we are not creating any user account but inviting the existing user to a particular org.

jain-prerna commented on src/saltext/vmware/states/vmc_org_users.py:
```diff
@@ -0,0 +1,417 @@
+                result=False,
+            )
+
+        return vmc_state._create_state_response(
+            name=name,
+            comment="Invited user {} successfully".format(username),
+            new_state=invited_org_user,
+            result=True,
+        )
```
It's not possible to know if the invite was already issued to the user.
and if the user is already added to the org and we keep invoking the invite API again it always return "202 accepted ".

We tried with both valid email and test email and  in both the cases while invoking invite API multiple times, it just returns "202 accepted".

waynew replied:
Is there no API function to call to see if the invite is pending?

jain-prerna replied:
I found an API that return invitation  when called with org_id and invitation_id 
https://{api_host}/csp/gateway/am/api/orgs/{orgId}/invitations/{userInvitationId}

but there are two challenges with this API as well
- i tired to invite multiple users who has valid vmware account and the one with test account using invitation API and then called  the above API, it  return only the test account in the response with available status but not the ither users having vmware account.
 And as per observation the valid user account does not require to accept invitation , they get added to org as the invitation triggers
- the API used to send invitation does not return anything in response but the http status, so there we would need another call to get invitation_id itself.

Can we discuss it on call today ?  i can show you the APIs available and their  responses

jain-prerna commented:
Hi @waynew , i have addressed all the comments here, Please let us know if any other changes are needed.

waynew reviewed (changes requested):
Still those two unresolved conversations that need to be addressed, otherwise this appears to work well enough :+1:

jain-prerna commented:
Hi @waynew , I have modified the invite function with below flow as per our discussion, Could you please review it

Adding user:
-Send request to API to add user
-Search for user - if found, return "Added user {username/email/whatever}"
- If not found, search for invite. If found "Invited user {username/email/whatever}"
- If not found, "Failed to add user {username/email/whatever}"

waynew commented on src/saltext/vmware/modules/vmc_org_users.py:
```diff
@@ -232,111 +232,238 @@ def invite(
-        'expiresAt': (Integer As Int64) (Optional)
-            The timestamp the role expires at (measured in number of seconds since 1/1/1970 UTC). Example:3609941597
+    custom_groups_ids
+        (Optional) List of unique IDs of Custom Groups. Can have a maximum of 15 custom group IDs.
```
**praise** love to see this called out here, that there's a max. That will be super helpful to users.

waynew commented on src/saltext/vmware/states/vmc_org_user.py:
```diff
@@ -0,0 +1,330 @@
+        - cert: /path/to/client/certificate
+
+.. warning::
+    It is recommended to pass the VMC authentication details using Pillars rather than specifying as plain text in SLS
```
**note** with #258 nearly merged, we should be able to drastically improve these modules. I'm pretty sure we won't even need to take the hostname/refresh_key/org_id anymore, instead, the config layout will look something like this:

```
saltext.vmware:
  vmc:
    host: 203.113.0.42
    refresh_key: abcd1234
    org_id: 1234abcd
```
alternatively:
```
saltext.vmware:
  profile_1:
    vmc:
      host: 203.113.0.42
      refresh_key: abcd1234
      org_id: 1234abcd
  profile_2:
    vmc:
      host: 203.113.0.43
      refresh_key: abcd1234
      org_id: 1234abcd
```

jain-prerna replied:
@waynew , we will take it as a separate PR to make change in all the modules

waynew reviewed (changes requested):
Here are a couple of things for our meeting - there might be some other changes that we want to make as well.

waynew commented on src/saltext/vmware/states/vmc_org_user.py:
```diff
@@ -0,0 +1,330 @@
+            comment="Failed to get users for given org : {}".format(org_users_list["error"]),
+            result=False,
+        )
+    org_user = None
+    for org_user in org_users_list.get("results"):
+        if username == org_user["user"].get("username"):
+            break
+        else:
+            org_user = None
```
**suggestion** rather than changing the org_user every time, this could be re-written to use the for/else construct:
```
for org_user in org_users_list["results"]:
    if username == org_user["name"].get("username"):  # will the user ever be missing a username?
        break
else:
    org_user = None
```

Or it could be re-written with a generator comprehension/expression:

```
org_user = next((user for user in org_users_list["results"] if user["user"].get("username") == username), None)
```

jain-prerna replied:
Updated

jain-prerna commented:
> Here are a couple of things for our meeting - there might be some other changes that we want to make as well.

Okay @waynew i am going through your comments and will address it.
Let's meet sync up today if any other change is required there

jain-prerna commented:
https://developer.vmware.com/apis/csp/csp-iam/latest/csp/gateway/am/api/orgs/orgId/users/search/get/

This API doc has changed recently and it has been marked deprecated and so i have to modify functions accordingly.

What i have found through VMC channels , this is the new POST method API  that will the purpose https://developer.vmware.com/apis/csp/csp-iam/latest/csp/gateway/am/api/orgs/orgId/users/search/post/

I am working on it ,  i was using this API in org_users module as a separate function "search" and in the "add" function

jain-prerna commented on src/saltext/vmware/modules/vmc_org_users.py:
```diff
@@ -431,22 +546,70 @@ def invite(
+            cert=cert,
+        )
+
+        if invitation_list:
+            for invitation in invitation_list:
+                if user_name == invitation["username"] and invitation["status"] == "AVAILABLE":
+                    log.info("Invited  %s successfully", user_name)
+                    result = invitation
+                    break
+                else:
+                    log.info("Failed to add %s", user_name)
+                    result = {
+                        vmc_constants.ERROR: "Failed to add {}".format(user_name, description)
+                    }
```
empty invidation_list should consider as failure, Handled it

jain-prerna replied:
Yes invidation_list is always going to be iterable, cross checked the API reponse

jain-prerna commented on src/saltext/vmware/states/vmc_org_user.py:
```diff
@@ -0,0 +1,330 @@
+.. code-block:: yaml
+
+    test_user@vmware.com:
+      vmc_org_user.invite:
```
Got it , Updated.

jain-prerna commented:
@waynew Addressed all the comments , tested all function and verified unit & integration tests.

pre-commit is failing after pushing code from local, it seems failure caused  after pulling the latest

waynew commented:
@jain-prerna looks like there are some failing tests - if you want to take a look before (my) tomorrow, I should be available in the morning.

jain-prerna commented:
> @jain-prerna looks like there are some failing tests - if you want to take a look before (my) tomorrow, I should be available in the morning.



> @jain-prerna looks like there are some failing tests - if you want to take a look before (my) tomorrow, I should be available in the morning.

Fixed it, all checks are green.

waynew reviewed (approved):
I had to update the state doc to match the state function name, and restored the esxi.py file, but I think this  ready now.
